src/lib/ipsec/esp.lua

```diff
@@ -168,10 +162,59 @@ function esp_v6_decrypt:decapsulate (p)
                    "flow_id=" .. tostring(self.ip:flow_label()) .. ", " ..
                    "reason='" .. reason .. "'";
       logger:log("Rejecting packet ("..info..")")
-      return false
+      -- RFC4303 is somewhat unclear on a) whether or not packets that look replayed are eligible to trigger
+      -- recyns, and b) what 'consecutive' means in "number of consecutive packets that fail authentication":
+      -- Consecutive in time? Consecutive in their seq_low's?
+      -- Assuming a) should be 'yes' since the to-be-resynced packets might have seq_low's that just happen
+      -- to fall inside the replay window (seq_lo-wise) and would look replayed, yet aren't.
+      -- Assuming b) means 'in time', since systematic loss could stall resync indefinitely.
+      proceed = false
+      self.decap_fail = self.decap_fail + 1
+      if self.decap_fail >= ESP_RESYNC_THRESHOLD then
+         local seq_high_tmp = seq_high
+         if seq_high >= 0 then -- We failed to decrypt in-place, undo the damage to recover the original ctext
+            gcm:encrypt(ctext_start, iv_start, seq_low, seq_high, ctext_start, ctext_length, ffi.new("uint8_t[?]", gcm.AUTH_SIZE))
+         else
+            seq_high_tmp = self:seq_high() -- use the last seq_high we saw if it looked replayed
+         end
+         self.decap_fail = 0 -- avoid immediate re-triggering if resync fails
+         seq_high_tmp = self:resync(p, seq_low, seq_high_tmp + 1, ESP_RESYNC_ATTEMPTS)
+         if seq_high_tmp >= 0 then
+            seq_high = seq_high_tmp
+            proceed = true -- resynced! the data has been decrypted in the process so we're ready to go
+         end
+      end
    end
+   if not proceed then return false end
```

Comment:
We could drop `proceed` at the cost of two `else return false`, yes.  Good catch.